File: app/src/main/java/com/qos/myapplication/MainActivity.java
package com.qos.myapplication;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.qos.myapplication.databinding.ActivityMainBinding;
import com.qos.myapplication.permissionmanager.RequestPermissions;
import com.qos.myapplication.tests.DeviceInformation;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    private RequestPermissions requestPermissions;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        requestPermissions = new RequestPermissions(this);
        SharedPreferences sharedPreferences = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE);
        boolean permissionsRequested = sharedPreferences.getBoolean("permissionsRequested", false);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        if(!permissionsRequested){
            requestPermissions.checkAndRequestPermission();
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putBoolean("permissionsRequested", true);
            editor.apply();
        }

        BottomNavigationView navView = findViewById(R.id.nav_view);
        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        AppBarConfiguration appBarConfiguration = new AppBarConfiguration.Builder(
                R.id.navigation_home, R.id.navigation_dashboard, R.id.navigation_notifications)
                .build();
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_activity_main);
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);
        NavigationUI.setupWithNavController(binding.navView, navController);
    }

}

File: app/src/main/java/com/qos/testnet/MainActivity.kt
package com.qos.testnet

import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.qos.testnet.databinding.ActivityMainBinding
import com.qos.testnet.permissionmanager.PermissionPreferences
import com.qos.testnet.permissionmanager.RequestPermissions
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
    private var binding: ActivityMainBinding? = null
    private var requestPermissions: RequestPermissions? = null
    private val permissionPreferences: PermissionPreferences by lazy { PermissionPreferences.getInstance() }
    private lateinit var mAuth: FirebaseAuth
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        mAuth = FirebaseAuth.getInstance()
        requestPermissions = RequestPermissions(this)

        CoroutineScope(Dispatchers.Main).launch {
            val permissionsRequested = permissionPreferences.getPermissionPreference(
                this@MainActivity,
                PermissionPreferences.PermissionPreferencesKeys.ASK_OPEN_PERMISSION
            )

            if (!permissionsRequested) {
                requestPermissions?.requestAllPermissionsDialog()

                permissionPreferences.savePermissionPreference(
                    this@MainActivity,
                    PermissionPreferences.PermissionPreferencesKeys.ASK_OPEN_PERMISSION, true
                )
            }
        }

        if(!requestPermissions?.hasWriteStoragePermissions()!! || !requestPermissions?.hasReadStoragePermissions()!!) {
            requestPermissions!!.requestWriteStoragePermissions()
            requestPermissions!!.requestReadStoragePermissions()
        }

        if (mAuth.currentUser == null) {
            singInAnonymously()
        } else {
            val user: FirebaseUser? = mAuth.currentUser
            user?.let {
                Log.d("MainActivity", "User already signed in, userId: ${it.uid}")
            }
        }

        lifecycleScope.launch {
            val usrId = permissionPreferences.getUserId(
                this@MainActivity,
                PermissionPreferences.PermissionPreferencesKeys.USER_ID
            )
            Log.d("PreferencesStored", "Collected user Id: $usrId")
        }

        val navView = findViewById<BottomNavigationView>(R.id.nav_view)
        val navController = Navigation.findNavController(this, R.id.nav_host_fragment_activity_main)

        val appBarConfiguration = AppBarConfiguration(
            setOf(
                R.id.navigation_home,
                R.id.navigation_dashboard,
                R.id.navigation_notifications
            )
        )

        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration)
        NavigationUI.setupWithNavController(navView, navController)
    }

    private fun singInAnonymously() {
        mAuth.signInAnonymously()
            .addOnCompleteListener(this) { task ->
                if (task.isSuccessful) {
                    val user: FirebaseUser? = mAuth.currentUser
                    user?.let {
                        lifecycleScope.launch {
                            permissionPreferences.saveUserId(
                                this@MainActivity,
                                PermissionPreferences.PermissionPreferencesKeys.USER_ID, it.uid
                            )
                            Log.d("MainActivity", "SingInAnonymously:success, user ID:  ${it.uid}")
                        }

                    }
                } else {
                    Log.w("MainActivity", "SingInAnonymously:failure", task.exception)
                    Toast.makeText(this@MainActivity, "Authentication failed", Toast.LENGTH_LONG)
                        .show()
                }
            }
    }
}
